### ISSUE TITLE
Assignability of overload sets

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#4480]
ty and all other type checkers currently treat `Source` as being assignable to `Target`, but the first overload seems like a violation of the target `read` signature?

```py
from typing import Any, Literal, Protocol, TypeIs, overload

class Source:
    @overload
    def read(self, key: Literal["value"]) -> object: ...
    @overload
    def read(self, key: str) -> Any: ...
    def read(self, key: str) -> Any:
        raise NotImplementedError

class Target(Protocol):
    def read(self, key: str) -> int: ...

def f(target: Target): ...

f(Source())
```

**Comment by sharkdp** [comment]
The [spec says](https://typing.python.org/en/latest/spec/callables.html#overloads) "If a callable B is overloaded with two or more signatures, it is [assignable](https://typing.python.org/en/latest/spec/glossary.html#term-assignable) to callable A if **at least one of the overloaded signatures** in B is assignable to A". But it seems to me like that should only apply to overloads with non-overlapping parameter types?

Related issue: https://github.com/astral-sh/ty/issues/103
